beets/util/artresizer.py

```diff
@@ -83,14 +84,50 @@ def pil_resize(maxwidth, path_in, path_out=None, quality=0):
             quality = -1
 
         im.save(util.py3_path(path_out), quality=quality)
-        return path_out
+        if max_filesize > 0:
+            # If maximum filesize is set, we attempt to lower the quality of
+            # jpeg conversion by a proportional amount, up to 3 attempts
+            # First, set the maximum quality to either provided, or 95
+            if quality > 0:
+                lower_qual = quality
+            else:
+                lower_qual = 95
+            for i in range(3):
+                # 3 attempts is an abitrary choice
+                filesize = os.stat(path_out).st_size
+                log.debug(u"PIL Pass {0} : Output size: {1}B", i, filesize)
+                if filesize <= max_filesize:
+                    return path_out
+                # Calculate percentage decrease required. This is approximate,
+                # as quality may only loosely correlate with filesize.
+                frac_diff = 1 - ((filesize - max_filesize) / max_filesize)
+                assert frac_diff < 1
+                lower_qual = lower_qual * frac_diff
+                # Restrict quality dropping below 10
+                if lower_qual < 10:
+                    lower_qual = 10
+                # Use optimize flag to improve filesize decrease
+                im.save(
+                    util.py3_path(path_out), quality=lower_qual, optimize=True
+                )
+            log.error(u"PIL Failed to resize file to below {0}B", max_filesize)
+            raise IOError
+
+        else:
+            return path_out
     except IOError:
         log.error(u"PIL cannot create thumbnail for '{0}'",
                   util.displayable_path(path_in))
         return path_in
+    except AssertionError:
+        log.error(
+            u"Percentage filesize decrease calculated incorrectly for '{0}'",
+            util.displayable_path(path_in),
+        )
+        return path_in
```

Comment:
I think this should be dropped: Either the `assert` truly has the purpose of an assertion, i.e. checks for programmer error, and should be impossible to happen, or it is an error that depends on runtime state. Then, just directly raise some error or `return path_in`, without the intermediate assertion.

Reply:
Agreed, I think this was left in from writing the algorithm when I was making some basic mathematical errors, but it serves no purpose anymore.